File: src/pipeline/detectors/c_incomplete_type.py
"""
Detector for C compilation errors for incomplete struct/type errors.
"""

import re
import typing as T
from src.pipeline.detectors.base import Detector
from src.pipeline.models import ErrorClue


class CIncompleteTypeDetector(Detector):
    """
    Detect C compilation errors for incomplete struct/type errors.

    Matches patterns like:
    - error: field 'orig_termios' has incomplete type
    - error: storage size of 'raw' isn't known
    - These usually indicate missing header files for struct definitions
    """

    PATTERNS = {
        # Refactored to avoid multiple lazy quantifiers - uses greedy matching within lines
        "missing_c_include": r"(?P<file_path>[a-zA-Z0-9_./\-]+\.c):\d+:\d+:[^\n]*error:[^\n]*(?:has incomplete type|storage size).*struct\s+(?P<struct_name>termios|winsize|stat|tm|sigaction|dirent)",
    }

    EXAMPLES = [
        (
            "dim.c:81:18: error: field 'orig_termios' has incomplete type\n   81 |   struct termios orig_termios;",
            {
                "clue_type": "missing_c_include",
                "confidence": 1.0,
                "context": {
                    "file_path": "dim.c",
                    "struct_name": "termios",
                },
            },
        ),
    ]

File: src/boil/detectors/c_incomplete_type.go
// Detector for C incomplete type errors.
package detectors

// CIncompleteTypeDetector detects C incomplete type errors
type CIncompleteTypeDetector struct {
	*BaseDetector
}

func NewCIncompleteTypeDetector() (*CIncompleteTypeDetector, error) {
	base, err := NewBaseDetector(
		"CIncompleteTypeDetector",
		100,
		map[string]string{
			"missing_c_include": `(?P<file_path>[a-zA-Z0-9_./\-]+\.c):\d+:\d+:[^\n]*error:[^\n]*(?:has incomplete type|storage size).*struct\s+(?P<struct_name>termios|winsize|stat|tm|sigaction|dirent)`,
		},
		[]DetectorExample{
			{
				Name:     "incomplete_type_termios",
				Input:    "term.c:15:3: error: variable has incomplete type 'struct termios'",
				ClueType: "missing_c_include",
				Context: map[string]string{
					"file_path":   "term.c",
					"struct_name": "termios",
				},
			},
			{
				Name:     "storage_size_winsize",
				Input:    "window.c:8:12: error: storage size of 'ws' isn't known struct winsize",
				ClueType: "missing_c_include",
				Context: map[string]string{
					"file_path":   "window.c",
					"struct_name": "winsize",
				},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &CIncompleteTypeDetector{base}, nil
}
